lib/icinga/clusterevents.cpp

```diff
@@ -1188,6 +1181,14 @@ Value ClusterEvents::ExecutedCommandAPIHandler(const MessageOrigin::Ptr& origin,
 		return Empty;
 	}
 
+	if (origin->FromZone && !origin->FromZone->CanAccessObject(checkable) &&
+		endpoint->GetName() != execution->Get("endpoint")) {
```

Comment:
This still has the issue that it won't work with additional satellites in between. Unfortunately the naming is a bit misleading: `origin->FromClient->GetEndpoint()` isn't the endpoint on which the message originated but the one from which the current node received it.

Without having tested this: I think `!origin->FromZone->CanAccessObject(Endpoint::GetByName(execution->Get("endpoint")))` (+ additional checks to ensure that the endpoint (still) exists) should do the trick. This should then also make the check for `!origin->FromZone->CanAccessObject(checkable)` obsolete.